hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java

```diff
@@ -542,8 +542,17 @@ protected PingInputStream(InputStream in) {
        */
       private void handleTimeout(SocketTimeoutException e, int waiting)
           throws IOException {
+        int timeout = 0;
+        if (rpcTimeout > 0) {
+          // effective rpc timeout is rounded up to multiple of pingInterval
+          // if pingInterval < rpcTimeout.
+          timeout = (doPing && pingInterval < rpcTimeout) ?
+              pingInterval : rpcTimeout;
+        } else {
+          timeout = pingInterval;
+        }
         if (shouldCloseConnection.get() || !running.get() ||
-            (0 < rpcTimeout && rpcTimeout <= waiting)) {
+            (timeout <= waiting)) {
```

Comment:
Is your changed code the same as soTimeout <= waiting?

Reply:
Yes, now I just basically follow the usage of rpcTimeout in the constructor of hadoop.ipc.Client$Connection class.
However, I think using soTimeout is not suitable (although the value is the same for now), because they should have different meaning.